File: PGProblemSolving/Stack,Queue/다리를지나는트럭_42583.cpp
//
//  다리를 지나는 트럭_42583.cpp
//  ProgrammersAlgorithmPractice
//
//  Created by MinKyeongTae on 07/09/2019.
//  Copyright © 2019 Min Kyeong Tae. All rights reserved.
//

// MARK: 다리를지나는트럭 풀이답안
#if 0
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int time=0, total=0, idx=0;
    queue<int> Q;
    
    while(idx < truck_weights.size()) {
        if(total+truck_weights[idx] <= weight) {
            Q.push(truck_weights[idx]);
            total+=truck_weights[idx];
            idx++;
        } else Q.push(0);
        
        if(Q.size() >= bridge_length) {
            total-=Q.front();
            Q.pop();
        }
        time++;
    }
    return time + bridge_length;
}
#endif

File: PGProblemSolving/Stack,Queue/다리를지나는트럭_42583.swift
//
//  다리를지나는트럭.swift
//  ProgrammersAlgorithmPractice
//
//  Created by MinKyeongTae on 13/09/2019.
//  Copyright © 2019 Min Kyeong Tae. All rights reserved.
//

// MARK: - 다리를 지나는 트럭_42583 : Queue Algorithm Problem

/*
 import Foundation

 func TrucksPassingBridge(_ bridge_length: Int, _ weight: Int, _ truck_weights: [Int]) -> Int {
 var bridge = [Int]()
 var truckWs = truck_weights
 var sum = 0
 var Ans = 0
 for i in truck_weights.indices {
     while true {
         // 진입하는 트럭의 무게허용 가능일 경우
         if sum + truckWs[i] <= weight {
             bridge.append(truckWs[i])
             sum += truckWs[i]
             Ans += 1
             break
         } else {
             // 트럭의 무게중량 초과 + 큐 길이가 찼을때
             if bridge.count >= bridge_length {
                 sum -= bridge[0]
                 bridge.removeFirst()
                 // 트럭의 무게중량 초과 + 큐 길이가 안 찼을때
             } else {
                 bridge.append(0)
                 Ans += 1
             }
         }
     }
 }

 Ans += bridge_length
 return Ans
 }
 */
